core/contracts/common/implementation/Lockable.sol

```diff
@@ -0,0 +1,53 @@
+pragma solidity ^0.6.0;
+
+
+// TODO:
+// - Conform comments to Natspec
+// - Add unit tests
+
+// Extends https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/ReentrancyGuard.sol
+// with view-method lock inspired by https://github.com/balancer-labs/balancer-core/blob/master/contracts/BPool.sol
+
+contract Lockable {
+    // TODO: I don't like this name but its an intuitive placeholder.
+    bool private _notEntered;
+
+    constructor() internal {
+        // Storing an initial non-zero value makes deployment a bit more
+        // expensive, but in exchange the refund on every call to nonReentrant
+        // will be lower in amount. Since refunds are capped to a percetange of
+        // the total transaction's gas, it is best to keep them low in cases
+        // like this one, to increase the likelihood of the full refund coming
+        // into effect.
+        _notEntered = true;
+    }
+
+    /**
+     * @dev Prevents a contract from calling itself, directly or indirectly.
+     * Calling a `nonReentrant` function from another `nonReentrant`
+     * function is not supported. It is possible to prevent this from happening
+     * by making the `nonReentrant` function external, and make it call a
+     * `private` function that does the actual work.
+     */
+    modifier nonReentrant() {
+        // On the first call to nonReentrant, _notEntered will be true
+        require(_notEntered, "ReentrancyGuard: reentrant call");
+
+        // Any calls to nonReentrant after this point will fail
+        _notEntered = false;
+
+        _;
+
+        // By storing the original value once again, a refund is triggered (see
+        // https://eips.ethereum.org/EIPS/eip-2200)
+        _notEntered = true;
+    }
```

Comment:
Would it help the bytecode size if we put all the logic in two (or just one) internal methods `_preEntranceCheck()` and `_postEntranceReset()` (the latter may not help/be necessary)? Obviously, it's messier, but it just wanted to try it in case it would slightly improve our bytecode issues (my worry is that the require call is taking a lot of bytecode and being copied around to every method.

Reply:
Wouldn't this have the same number of require statements per function call?

Reply:
The *logic* would be the same. But the way Solidity organizes the bytecode may be different. For instance, I think modifiers generally work this way:
```
function 1 start
modifier bytecode1 ( pre _; )
function 1 logic bytecode
modifier bytecode2 ( post _; )
function 1 end

function 2 start
modifier bytecode1 ( pre _; )
function 2 logic bytecode
modifier bytecode2 ( post _; )
function 2 end

...
```

The modifier bytecode would just be pasted into every function.
If you were to have the modifier call a function, it might look like this:
```
function 1 start
JUMP to modifier bytecode1
function 1 logic bytecode
JUMP to modifier bytecode2
function 1 end

function 1 start
JUMP to modifier bytecode1
function 1 logic bytecode
JUMP to modifier bytecode2
function 1 end

...

modifier bytecode1 (pre _; )
modifier bytecode2 (post _; )
```

In the latter, there's only one instruction to run each part of the modifier bytecode -- jumping to the function that the modifier calls. In the former, the entire modifier bytecode will be pasted for each function, bloating the bytecode. All the opcodes and data that make up that require statement may only be in one place in the bytecode rather than copied into every function. We've established in the past that require statements have a pretty big impact on the bytecode (especially when they have a message), so I thought eliminating the repetition may help -- although I'm not 100% sure of how solidity organizes this. It could require you changing the number of `runs` in the optimizer to see the impact.

Reply:
Ok this would help, but the overall bytecode impact would be small, likely < 1000 bytes. We'll run into the bytecode limit either way once we add back require messages. I think we should think bigger picture about how to organize the code to save bytecode.

I do think this change would help some and I'll test it out.

Reply:
This saved ~800 bytes, great idea @mrice32 !